### DISCUSSION TITLE
kube-vip as a lb?

### DISCUSSION THREAD

**Discussion by ndrwstn** [techno-tim/k3s-ansible#174] [Questions]
In his video on this topic, @timothystewart6 said he couldn't get kube-vip to work as a load balancer, which is why we're still using metallb. I didn't think much of this at the time--nor did I mind, as I like metallb--but after getting this repo working with kube-vip using BGP, I just found out that a single node can't use BGP in two deployments (e.g. kube-vip and metallb) as it leads to 'peer flapping', which I need to solve.

The simple solution to this is likely to just use kube-vip as a load balancer, but before I start down that rabbit-hole, I wanted to see if there was some Very Good Reason why kube-vip didn't work in the first place? Anyone have any ideas?

**Comment by timothystewart6** [comment]
It might have been a bug or a configuration issue at the time.  I am also considering making metal-lb optional.  https://github.com/techno-tim/k3s-ansible/issues/172

**Reply by taboneclayton** [reply]
Sorry for the long-winded answer. This reply is meant to serve as a guideline for anyone who wants to go the kube-vip route for both control-plane and as a service load balancer.

This is possible with this playbook with minor modifications. Tested locally. I might present a PR eventually but given discussions in #172 and #116 I am not sure whether there is interest in adding a LB alternative to this playbook. For anyone interested though, here's a summary of the required changes:
- disable MetalLB by physically commenting out the task from the repo or wait for #172 to be completed and then disable MetalLB
- set `svc_enable` to true in [vip.yaml.j2](https://github.com/techno-tim/k3s-ansible/blob/master/roles/k3s/master/templates/vip.yaml.j2#L43) to configure kube-vip to be used also as a service load balancer
- Install kube-vip-cloud-controller by following the [official instructions](https://kube-vip.io/docs/usage/cloud-provider/#install-the-kube-vip-cloud-provider)

The kube-vip-cloud-controller is an IP Address Management (IPAM) solution for on-premise installations. Kubernetes would usually request a public IP from a cloud provider such as AWS, Google Cloud, Azure, etc. Within an on-premise installation, kube-vip requires the kube-vip-cloud-controller to be able to provide an IP when kube-vip requests one. It is important to create the `kubevip` ConfigMap in the `kube-system` namespace to configure the range of IPs which will be used by kube-vip-cloud-controller when handing out IPs to kube-vip. Again follow [official instructions](https://kube-vip.io/docs/usage/cloud-provider/#the-kube-vip-cloud-provider-configmap) for this.

An optional task can be added to the playbook to install the kube-vip-cloud-controller manifest as part of the playbook. Alternatively, the kube-vip-cloud-controller manifest can be installed manually after running the k3s-ansible playbook with MetalLB disabled.

Nowadays GitOps has become a prevalent way to install/maintain clusters. So my personal preference would be run the playbook with MetalLB disabled and then leverage Flux/ArgoCD to install the kube-vip-cloud-controller manifest. This would allow us to keep this playbook as simple as possible (reference to @timothystewart6's comment in #192) while still supporting kube-vip as an LB. The only thing to keep in mind is that the `svc_enable` within the kube-vip DaemonSet would needs to also be configured outside of this Ansible playbook.

**Reply by timothystewart6** [reply]
Hey, agree with all of these statements.  I would love to give `kube-vip` some more flexibility as well as the option to exclude MetalLB. I too fee like GitOps is the way to go vs. installing anything via ansible see https://github.com/techno-tim/k3s-ansible/pull/192#issuecomment-1376588631

**Reply by rr4444** [reply]
Don't disagree, but then whats the point of this vs k3s-io/k3s-ansible ?

Is that one too  stale? Maybe I missing repo functionality knowledge and so am missing the point....

**Reply by timothystewart6** [reply]
No clue about the differences, last I checked that didn't work nor was supported.  This ones creates an HA cluster with HA service load balancer and HA k8s API.

**Reply by rr4444** [reply]
Def def, totally agree. So:

1. if the official k3s ansible repo is verging on abandoned
2. and this repo's ansible element is instead changed to be only for k3s deployment
3. meaning the kubevip +/- metal-lb combo moves from ansible to k8s yaml deployment

then this repo becomes THE community maintained ansible k3s deployment repo (plus bells and whistles)

That's still a very valid and valuable FOSS role. 

Or am I misunderstanding things?

Frankly, I am trying to work out the most sensible k8s + kubevip + kubevip cloud controller combo. And this repo feels nicely alive to have legs for the future.

**Reply by timothystewart6** [reply]
That's exactly right.  I couldn't find one that worked, so I forked and built this instead 😄
